cas-server-support-osf/src/main/java/io/cos/cas/adaptors/postgres/handlers/OpenScienceFrameworkInstitutionHandler.java

```diff
@@ -102,7 +105,12 @@ public String findInstitutionLogoutUrlById(final String id) {
         for (final OpenScienceFrameworkInstitution institution: institutionList) {
             final DelegationProtocol delegationProtocol = institution.getDelegationProtocol();
             if (DelegationProtocol.SAML_SHIB.equals(delegationProtocol)) {
-                institutionLoginUrlMap.put(institution.getLoginUrl() + "&target=" + target, institution.getName());
+                // With shared SSO between institutions such as Brown University and The Policy Lab, the login URL is
+                // the same for both. Must add "#" with institution ID to enforce key uniqueness.
+                institutionLoginUrlMap.put(
+                        institution.getLoginUrl() + "&target=" + target + '#' + institution.getId(),
+                        institution.getName()
+                );
```

Comment:
The initial solution is to only add the fragment with institution ID when the key has been put into the map.

```java
String loginUrlAsKey = institution.getLoginUrl() + "&target=" + target;
if (institutionLoginUrlMap.containsKey(loginUrlAsKey)) {
    loginUrlAsKey = loginUrlAsKey + '#' + institution.getId();
}
institutionLoginUrlMap.put(loginUrlAsKey, institution.getName());
```

However, I decide to apply this to all institutions for a more general approach.

Reply:
And a fun fact, java doesn't like string `"#"` and prefer character `'#'` when concatenating string. In addition, no need to use `StringBuilder` since it will be used by the compiler automatically. And it is the `StringBuilder` who doesn't like the one-character string lol ...